client/packages/requisitions/src/ResponseRequisition/ListView/ProgramRequisitionOptions.tsx

```diff
@@ -0,0 +1,191 @@
+import React, { useEffect, useState } from 'react';
+import {
+  Autocomplete,
+  AutocompleteProps,
+  ButtonWithIcon,
+  Grid,
+  PlusCircleIcon,
+  Typography,
+  useTranslation,
+} from '@openmsupply-client/common';
+import { getNameOptionRenderer } from '@openmsupply-client/system';
+
+import { CustomerProgramSettingsFragment } from '../api';
+import { NewRequisitionType } from '../../types';
+
+export interface NewProgramRequisition {
+  type: NewRequisitionType.Program;
+  programOrderTypeId: string;
+  otherPartyId: string;
+  periodId: string;
+}
+
+type Common<T> = Pick<
+  AutocompleteProps<T>,
+  'options' | 'value' | 'disabled' | 'renderOption' | 'getOptionDisabled'
+> & {
+  label: string;
+  set: (value: T | null) => void;
+  labelNoOptions?: string;
+};
+
+const useProgramRequisitionOptions = (
+  programSettings: CustomerProgramSettingsFragment[]
+) => {
+  const t = useTranslation();
+  type ProgramSetting = CustomerProgramSettingsFragment;
+  type CustomerAndOrderTypes =
+    CustomerProgramSettingsFragment['customerAndOrderTypes'][number];
+  type OrderType = CustomerAndOrderTypes['orderTypes'][number];
+  type Customer = CustomerAndOrderTypes['customer'];
+  type Period = OrderType['availablePeriods'][number];
+
+  const [program, setProgram] = useState<ProgramSetting | null>(null);
+  const [orderType, setOrderType] = useState<OrderType | null>(null);
+  const [customer, setCustomer] = useState<Customer | null>(null);
+  const [period, setPeriod] = useState<Period | null>(null);
+
+  useEffect(() => {
+    setOrderType(null);
+    setCustomer(null);
+  }, [program]);
+
+  useEffect(() => {
+    setPeriod(null);
+  }, [orderType]);
+
+  const allOptions: {
+    programs: Common<ProgramSetting>;
+    orderTypes: Common<OrderType>;
+    customers: Common<Customer>;
+    periods: Common<Period>;
+  } = {
+    programs: {
+      options: programSettings,
+      value: program,
+      set: setProgram,
+      label: t('label.program'),
+      disabled: false,
+    },
+    orderTypes: {
+      options: program?.customerAndOrderTypes.flatMap(c => c.orderTypes) || [],
+      value: orderType,
+      set: setOrderType,
+      disabled: program === null,
+      labelNoOptions: t('messages.not-configured'),
+      label: t('label.order-type'),
```

Comment:
Whenever I see useEffects I always review deeply cause they're a common performance mishap! There is an amazing bit of the react docs that basically goes through all the common do's and don'ts of useEffect - this one applies here https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state

You can get the same behaviour by writing event handlers like below (it's tempting to have the 4 variables in one state object and use useReducer but it's probably overkill?):

(**Don't accept these changes with the console.log in there lol**)
```suggestion
let count = 0;
const useProgramRequisitionOptions = (
  programSettings: CustomerProgramSettingsFragment[]
) => {
  const t = useTranslation();
  type ProgramSetting = CustomerProgramSettingsFragment;
  type CustomerAndOrderTypes =
    CustomerProgramSettingsFragment['customerAndOrderTypes'][number];
  type OrderType = CustomerAndOrderTypes['orderTypes'][number];
  type Customer = CustomerAndOrderTypes['customer'];
  type Period = OrderType['availablePeriods'][number];

  const [program, setProgram] = useState<ProgramSetting | null>(null);
  const [orderType, setOrderType] = useState<OrderType | null>(null);
  const [customer, setCustomer] = useState<Customer | null>(null);
  const [period, setPeriod] = useState<Period | null>(null);
  console.log('render', count++);

  const handleSetProgram = (value: ProgramSetting | null) => {
    setProgram(value);
    setOrderType(null);
    setCustomer(null);
    setPeriod(null);
  };

  const handleSetOrderType = (value: OrderType | null) => {
    setOrderType(value);
    setPeriod(null);
  };

  const allOptions: {
    programs: Common<ProgramSetting>;
    orderTypes: Common<OrderType>;
    customers: Common<Customer>;
    periods: Common<Period>;
  } = {
    programs: {
      options: programSettings,
      value: program,
      set: handleSetProgram,
      label: t('label.program'),
      disabled: false,
    },
    orderTypes: {
      options: program?.customerAndOrderTypes.flatMap(c => c.orderTypes) || [],
      value: orderType,
      set: handleSetOrderType,
      disabled: program === null,
      labelNoOptions: t('messages.not-configured'),
      label: t('label.order-type'),
```

A certain path I was testing:

1. Set all 4 ddl values
2. Clear the Program (this would cause 2 useEffects to be triggered, one of which may retrigger the second again)

With the event handlers the renders you get is 1 (going from 5 to 6):
![image](https://github.com/user-attachments/assets/5d657e42-d570-45f9-8cc5-bf9177af366f)
With the original useEffects it'd go from 6 to 9. So this should save some performance!

Reply:
For what it's worth, a good simple rule for useEffect that I live by and got from [react.gg](react.gg):

> If a side effect is synchronising your component with some external system, put that side effect in useEffect. 

There is sparingly any other reason to useEffect, and almost all of our synchronising is fetch requests handled by useQuery, so it's surprising our code has >250 useEffects (could be a significant source of our performance issues and odd value jumping/rerender bugs!)

Reply:
Agree, useEffect can cause a lot of problems later on as well [68d62d7](https://github.com/msupply-foundation/open-msupply/pull/5220/commits/68d62d700fc7f226891ad6eef2a707cd10710b46)